src/main/java/fi/aalto/cs/apluscourses/dal/TokenAuthentication.java

```diff
@@ -0,0 +1,55 @@
+package fi.aalto.cs.apluscourses.dal;
+
+import fi.aalto.cs.apluscourses.model.Authentication;
+import java.util.Arrays;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public abstract class TokenAuthentication implements Authentication {
```

Comment:
I wonder if this could just be a concrete class with the `APlusTokenAuthentication` implementation. As far as I can tell, we use either `APlusTokenAuthentication` or `Authentication` everywhere (even in tests), so this seems a bit like adding an abstract `Mammal` class between `Animal` and `Dog`; it makes sense but is it needed

Reply:
Just trying to apply single-responsibility principle here. I think `AuthenticationViewModel` (and hence the corresponding view, too) refer to this class (via its factory) instead of `APlusTokenAuthentication`. That way, if we have to change the implementation of `APlusTokenAuthentication` (for example, because of API changes) we can restrict more classes out of those that are potentially affected by that change.

But if you want, I can change this as you suggested.

Reply:
don't sag that easily :wink:

Reply:
I think the public interface of `TokenAuthentication` already provides the possibility to change the implementation without changing other classes, as the new implementation just needs to keep the same public interface. Also, the factory can return some other instance of `TokenAuthentication` if needed, even if `TokenAuthentication` itself already provides an implementation.

But this is anyways a minor thing and I've approved as-is, so your call :smile: